google/cloud/forseti/scanner/audit/external_project_access_rules_engine.py

```diff
@@ -0,0 +1,253 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Rules engine for external project access."""
+from collections import namedtuple
+import itertools
+import re
+
+from google.cloud.forseti.common.gcp_type import resource as resource_mod
+from google.cloud.forseti.common.gcp_type import resource_util
+from google.cloud.forseti.common.util import logger
+from google.cloud.forseti.scanner.audit import errors as audit_errors
+from google.cloud.forseti.scanner.audit import base_rules_engine as bre
+
+LOGGER = logger.get_logger(__name__)
+
+
+class ExternalProjectAccessRulesEngine(bre.BaseRulesEngine):
+    """Rules engine for External Project Access."""
+
+    def __init__(self, rules_file_path, snapshot_timestamp=None):
+        """Initialize.
+
+        Args:
+            rules_file_path (str): file location of rules
+            snapshot_timestamp (str): snapshot timestamp. Defaults to None.
+                If set, this will be the snapshot timestamp
+                used in the engine.
+        """
+        super(ExternalProjectAccessRulesEngine,
+              self).__init__(rules_file_path=rules_file_path)
+        self.rule_book = None
+
+    def build_rule_book(self, global_configs):
+        """Build ExternalProjectAccess rule book from the rules definition file.
+
+        Args:
+            global_configs (dict): Inventory configurations.
+        """
+        self.rule_book = (
+            ExternalProjectAccessRuleBook(
+                global_configs,
+                self._load_rule_definitions()))
+
+    # TODO: The naming is confusing and needs to be fixed in all scanners.
+    def find_policy_violations(self, user_email,
+                               project_ancestry,
+                               global_configs=None,
+                               force_rebuild=False):
+        """Determine whether project ancestry violates rules.
+
+        Args:
+            user_email (str): The user's e-mail
+            project_ancestry (list): List of ancestries which turn
+                                     out to a list of resources.
+            global_configs (dict): Inventory configurations. Necessary
+                                   if rebuilding a rule book.
+            force_rebuild (bool): Force the rebuild of the rule book
+
+        Returns:
+             generator: A generator of rule violations.
+        """
+
+        violations = itertools.chain()
+        if self.rule_book is None or force_rebuild:
+            self.build_rule_book(global_configs)
+        # If the ancestor found in the rule is not in the project_ancestry
+        # ancestors, then this is a rule violation
+        rules_violated = []
+        rules_not_violated = []
+        for ancestor, rule in self.rule_book.resource_rules_map.iteritems():
+            if ancestor not in project_ancestry:
```

Comment:
This is not a hard and fast rule, but there is an established pattern within our rule engines is that the exact logic for evaluation of a violation should happen inside each rule object.  This method here should be to:
1. build the rule book
2. `self.rule_book.find_violations(user_to_project_ancestry_map)`

Then in `rule book.find_violations()`, we do:
1. create `violations = []`
1. loop through each user in `user_to_project_ancestries_map`
1. `rule.find_violation(user, project_ancestries)`
1. append the returned violation object to `violations` list

Then in `rule.find_violation()`:
1. loop through the project_ancestries, and compare it with the rule's allowed organization.
1. if match, return violation object
1. else return None